hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-router/src/main/java/org/apache/hadoop/yarn/server/router/clientrm/FederationClientInterceptor.java

```diff
@@ -870,19 +870,88 @@ public ReservationDeleteResponse deleteReservation(
   @Override
   public GetNodesToLabelsResponse getNodeToLabels(
       GetNodesToLabelsRequest request) throws YarnException, IOException {
-    throw new NotImplementedException("Code is not implemented");
+    if (request == null) {
+      routerMetrics.incrNodeToLabelsFailedRetrieved();
+      RouterServerUtil.logAndThrowException("Missing getNodesToLabels request.", null);
+    }
+    long startTime = clock.getTime();
+    Map<SubClusterId, SubClusterInfo> subClusters =
+        federationFacade.getSubClusters(true);
```

Comment:
It looks like we do this pattern a bunch of time.
Maybe we can generalizer this with some lambda passed as a parameter?

Reply:
Is it possible to design a generic function like this?
```
  private <R> Collection<R> invokeConcurrent(
          Boolean filterInactiveSubClusters, ClientMethod request, Class<R> clazz)
          throws YarnException, RuntimeException {
    Map<SubClusterId, SubClusterInfo> subClusters =
            federationFacade.getSubClusters(filterInactiveSubClusters);
    return subClusters.keySet().stream().map(subClusterId -> {
      try {
        ApplicationClientProtocol protocol =
                getClientRMProxyForSubCluster(subClusterId);
        Method method = ApplicationClientProtocol.class
                .getMethod(request.getMethodName(), request.getTypes());
        return clazz.cast(method.invoke(protocol, request.getParams()));
      } catch (YarnException | NoSuchMethodException | IllegalAccessException |
               InvocationTargetException ex) {
        throw new RuntimeException(ex);
      }
    }).collect(Collectors.toList());
  }
```